ngraph_bridge/ngraph_encapsulate_clusters.h

```diff
@@ -35,10 +39,107 @@ typedef std::map<std::string, std::vector<int>> ShapeHintMap;
+// Cannot be copied/moved or reset
+class Encapsulator {
+ public:
+  Encapsulator(Graph* g);
+  // Populate ClusterManager with the subgraphs for each potential encapsulate
+  Status AnalysisPass();
+  // Perform the actual graph surgery
+  Status RewritePass(
+      FunctionDefLibrary* fdeflib, int graph_id,
+      const std::unordered_map<std::string, std::string>& device_config);
+  Status NewClusterIds(std::set<int>& result);
+
+  Encapsulator(const Encapsulator&) = delete;
+  Encapsulator(Encapsulator&&) = delete;
+  Encapsulator& operator=(const Encapsulator&) = delete;
+  Encapsulator& operator=(Encapsulator&&) = delete;
+
+ private:
+  Graph* graph;
+  // boolean to indicate if analysis has been done
+  // If not rewritepass should not be called
+  bool analysis_done;
+  // boolean to indicate that rewrite is done;
+  bool rewrite_done;
+  // A map from cluster indices to the expected device name for nodes
+  // in that cluster.
+  std::map<int, std::string> device_name_map;
+
+  // We *should* eventually have a way of monitoring the device and the backend
+  // together
+  std::map<int, std::string> backend_name_map;
+
+  // As we build the graph we will be tracking the.. TODO(amprocte): finish
+  // this comment.
+  std::map<std::tuple<int, int>, std::tuple<int, int>> output_remap_map;
+  std::map<std::tuple<int, int, int>, int> input_remap_map;
+  std::map<std::tuple<int, std::string, int>, string> input_rename_map;
+
+  // A map from cluster indices to a vector of input data types.
+  std::map<int, std::vector<std::tuple<int, int, DataType>>> cluster_input_map;
+  // A map from cluster indices to a vector of output data types.
+  std::map<int, std::vector<DataType>> cluster_output_dt_map;
+
+  // A map from cluster indices to corresponding NGraphEncapsulate nodes.
+  std::map<int, Node*> cluster_node_map;
+
+  std::set<int> cluster_indices_for_this_graph;
+
+  static void AddInput(NodeDef* dst, StringPiece src_name, int src_slot);
+};
+
+// Translates TF subgraph to ng function then compiles it
+Status PerformAOTOnEncapsulates(Graph* graph, const AOTInfo& aot_info);
+
+std::string HintAsString(ShapeHintMap single_hint);
+
+// Given a node, partialshape info from TF (present in the .pb itself) and a
+// shape hint, combine all that information
+PartialShape CombineNodeInfoAndHint(Node* node,
+                                    PartialShape partial_shape_from_node,
+                                    const ShapeHintMap& single_hint);
+
+// Given a TF graph, it scans it for inputs and finds what TF is saying about
+// their shapes (in the .pb itself)
+// Creates a map between input node names and PartialShape information we get
+// from the TF graph
+std::map<std::string, PartialShape> GetShapesFromTFInputnodes(
+    Graph* graph, const string& input_node_type);
+
+// Given an encapsulate node, and the input shapes,
+// performs TranslateGraph and returns an ng function and a signature
+Status PerformTranslation(Node* node,
```

Comment:
This function should be taking in a TFGraph rather than a Node. Can be addressed in some other PR

Reply:
This takes a node, which is expected to be an encapsulate (it throws an error otherwise). Given that it is an encapsulate, it finds its corresponding sub graph in the clustermanager and translates that.

Not the best API for a function as you mentioned. I could extract from the clustermanager beforehand and pass that to the function. 'll probably simplify it in the next PRs.